### ISSUE TITLE
to_df very slow on large datasets

### ISSUE THREAD

**Issue by bezdomniy** [sassoftware/saspy#298]
**Describe the bug**
I am using saspy over IOM to read large (~4gb) sas files into a pandas dataset. I am surprised that reading the whole dataset still takes over an hour. One issue that I found more surprising is that when I try to read a smaller dataset and benchmark the performance of to_df, to_df_CSV, and to_df_DISK, it turns out that to_df is much slower than the other 2.

For example to_df takes 401 seconds, to_df_CSV takes 108 seconds and to_df_DISK takes 158 seconds.

When I monitor the resource usage during to_df running, I find that the majority of the time the memory usage doesn't grow at all, and there is high cpu usage from the SAS 9.4 process. I am not familiar with how the library works, could it be something to do with the sas statements submitted here: https://github.com/sassoftware/saspy/blob/42e0e4e91aba7cc9d80198b746a8042a816898a0/saspy/sasioiom.py#L1609 ?

**Expected behavior**
I would expect that to_df should be the fastest, since according to the documentation it should be streaming through memory.

Could it be some kind of resource issue with how my SAS is configured? Could you suggest possible troubleshooting I could do?

**Desktop (please complete the following information):**
 - OS: Windows
 - SAS Version: 9.4M4
 - SASPy Version: 3.2.0

**Additional context**
Data is stored in sas datasets on the host's filesystem.

**Comment by tomweber-sas** [comment]
Hey @bezdomniy, there's some info on this here: https://sassoftware.github.io/saspy/advanced-topics.html#slow-performance-loading-sas-data-into-a-pandas-dataframe-to-df-sd2df

When I first wrote this, sd2df was the only version there was. You are correct that it streams data over and builds the dataframe in memory. It was pretty simple at the time. When trying to bring over larger datasets, it turned out to become very slow. This was all  on the python side, trying to cache lists of lists of data and build and concatenate dataframes from them. I reworked the algorithm for that to get it better, but after a point, it just consumed too much memory and cpu trying to build the dataframe in memory. FWIW, it is the fastest for smaller data, as there's no overhead that the other two have,

But, for bigger data, the CSV method is what I came up with, as csv is quite common for python to process. The idea was to have SAS export the dataset as a .csv file and use Pandas read_csv, which eliminates the memory and cpu thrashing of trying to build this on the fly. The read_csv method is a much faster way to create the dataframe; reading the file from disk. Proc Export in SAS (that creates the csv), doesn't support some delimiters, and embedded newlines and things like that well, so I ended up working around that problem by creating the DISK version. It used the original streaming method, where I can have direct control of the data, but instead of trying to build the DF in memory as I go, I just write it to disk to use the read_csv method of pandas to create the DF, like the CSV version.

What you saw, number wise, is about right. For larger files, CSV performs the best, followed closely by DISK, and then far behind by MEMORY. 

You are using IOM Remotely? SAS Workspace server is on some other machine? So, there's obviously the network speed in that situation, but there are the other factors too, on you PC side. So, I have a few ways to break this up and look at some of these components individually to get a better idea of where the time is. 

Fist thing, is no need to run any experiments with the full 4G, as 1G is big enough to show the proportionate behavior, and will waste 3/4 less of your time :) So, simply setting obs= to 1/4 the rows will let you 'shrink' the dataset.

I have a 4+G dataset and I have an IOM conection from my Linux, to my PC, so it's remote.
```
>>> sas
Access Method         = IOM
SAS Config name       = iomjwin
SAS Config file       = /opt/tom/github/saspy/saspy/sascfg_personal.py
WORK Path             = C:\Users\sastpw\AppData\Local\Temp\SAS Temporary Files\_TD43748_d10a626_\Prc2\
SAS Version           = 9.04.01M4P11142016
SASPy Version         = 3.3.6
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = cp1252
SAS process Pid value = 43748

>>> x = sas.sasdata('saspy','x')
>>> x.obs()
5000000
>>> sas.file_info(r'c:\Public\saspy.sas7bdat')
{'Create Time': '31Oct2017:16:36:35', 'File Size (bytes)': '4608147456', 'LRECL': '32767', 'Last Modified': '30Oct2017:15:38:42', 'Filename': 'c:\\Public\\saspy.sas7bdat', 'RECFM': 'V'}

>>> x
Libref  = x
Table   = saspy
Dsopts  = {}
Results = Pandas

>>> x.dsopts={'obs': 1000000}
>>> x
Libref  = x
Table   = saspy
Dsopts  = {'obs': 1000000}
Results = Pandas

>>> x.obs()
1000000
>>>
```

So, now I have 1/5 of the data to play with from the same 4G data set. The next thing is that with the CSV and DISK versions (no need to use MEMORY as there's no way this will perform on big data; that's a given), both allow you to write the data to a file location of your choice, and keep it instead of deleting it. You can observe how long it takes to write this file on your pc. And, you can also then just run the read_csv() by itself on that file to see how long that part takes.

One thing about IOM, as opposed to STDIO, is that the data to be streamed over has to first be written to a file on the SAS server side. In STDIO, I just open a socket and stream it directly over as it's read, but IOM I have to 'read a file' as the meas to stream it over. So, there is the extra overhead of reading the file in and writing it to a file to then be streamed over. You will see a delay when you run this before you start seeing the local file on your PC start to be populated. That delay is this part of the procedure. 
```
import time
>>> time.localtime(); df = x.to_df_DISK(tempfile='/opt/tom/saspy.5.csv', tempkeep=True); time.localtime()
time.struct_time(tm_year=2020, tm_mon=5, tm_mday=28, tm_hour=10, tm_min=37, tm_sec=24, tm_wday=3, tm_yday=149, tm_isdst=1)
time.struct_time(tm_year=2020, tm_mon=5, tm_mday=28, tm_hour=10, tm_min=40, tm_sec=25, tm_wday=3, tm_yday=149, tm_isdst=1)
>>>
>>> df.shape
(1000000, 115)
>>>

```
You can see from this, it took about 3 minutes to do the following:
1) read the SAS dataset and write it to a file on the SAS side in a format for read_csv later.
2) stream the data over to client and write it to to a local file
3) create the DF from the local file via pandas read_csv, and then do any necessary conversion if necessary.

You should be able to observe each of these 3 phases, simply by observing the local file. I see it get created, then no data gets written to it until step 1 completes. Once the file is 'full', step 2 is done. And when the sd2df_DISK finished, step 3 is over. 

I'm running this from linux, so observing the file is straight forward, in the end:
tom64-5> ll /opt/tom/saspy.5.csv
-rw-r--r-- 1 sastpw r&d 1764008703 May 28 10:39 /opt/tom/saspy.5.csv

Oh, and that line you  referenced in the source. That step is just defining a view, which doesn't take any time, really, and that part of the code is just getting the metadata about the SAS dataset so that I know what the columns are and then know how to convert the data to be able to create the data frame. There's no significant time being spent in that step.

I hope this all makes sense, and you can try this out and see what you observe. Once you see how much time is being taken be each step, maybe that will shed some light on it. 

I ran the full 4+G transfer next, as 1/5 took 3 min, so I'm guessing ~15 min for the whole 4G. Here's what I saw.

First, I see the file created, but nothing written while the SAS dataset is being converted to a file on the server side.

```
>>> time.localtime(); df = x.to_df_DISK(tempfile='/opt/tom/saspy.5.csv', tempkeep=True); time.localtime()
time.struct_time(tm_year=2020, tm_mon=5, tm_mday=28, tm_hour=10, tm_min=58, tm_sec=23, tm_wday=3, tm_yday=149, tm_isdst=1)

watchg the file. took 
tom64-5> ll /opt/tom/saspy.5.csv; date
-rw-r--r-- 1 sastpw r&d 0 May 28 10:58 /opt/tom/saspy.5.csv
Thu May 28 11:04:05 EDT 2020
tom64-5> ll /opt/tom/saspy.5.csv; date
-rw-r--r-- 1 sastpw r&d 0 May 28 10:58 /opt/tom/saspy.5.csv
Thu May 28 11:04:06 EDT 2020
tom64-5> ll /opt/tom/saspy.5.csv; date
-rw-r--r-- 1 sastpw r&d 9998691 May 28 11:04 /opt/tom/saspy.5.csv
Thu May 28 11:04:06 EDT 2020
tom64-5> ll /opt/tom/saspy.5.csv; date
-rw-r--r-- 1 sastpw r&d 19997235 May 28 11:04 /opt/tom/saspy.5.csv
Thu May 28 11:04:06 EDT 2020

Took about 5+ min to write the file out on server side. I was guess 1 min for
1/5 so, 5 min seems right. It's just my PC so not the best I/O or performance. 
Then streaming the data over, followed by creating the DF.

tom64-5> ll /opt/tom/saspy.5.csv; date
-rw-r--r-- 1 sastpw r&d 8820029735 May 28 11:08 /opt/tom/saspy.5.csv
Thu May 28 11:08:09 EDT 2020
tom64-5> ll /opt/tom/saspy.5.csv; date
-rw-r--r-- 1 sastpw r&d 8820033256 May 28 11:08 /opt/tom/saspy.5.csv
Thu May 28 11:08:09 EDT 2020
tom64-5> ll /opt/tom/saspy.5.csv; date
-rw-r--r-- 1 sastpw r&d 8820033256 May 28 11:08 /opt/tom/saspy.5.csv
Thu May 28 11:08:09 EDT 2020

That's the final size, so 4 min to stream it over. That's 9 min so far. And then
the DF creation via pandas: total time minus the 9.x min for the first 2 steps.
While it's running, I took a look at that process:

  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND
16935 sastpw    20   0 8965m 6.3g 1580 R 33.6 81.9   9:08.86 python3.5


At 11:15, which is 7+ more minutes of creating the dataframe, it failed w/ a memory
exception. It hadn't finished, so I don't really know how long it would have take if I
could have created a DF that big. I still have the input file, so I can try again to see
if it could work. Or just edit the file to shrink it to see how long it take for a bit
smaller;  keep trying till it works and extrapolate based upon the size that could
finally work.

>>> time.localtime(); df = x.to_df_DISK(tempfile='/opt/tom/saspy.5.csv', tempkeep=True); time.localtime()
time.struct_time(tm_year=2020, tm_mon=5, tm_mday=28, tm_hour=10, tm_min=58, tm_sec=23, tm_wday=3, tm_yday=149, tm_isdst=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/tom/github/saspy/saspy/sasdata.py", line 1148, in to_df_DISK
    return self.to_df(method='DISK', tempfile=tempfile, tempkeep=tempkeep, **kwargs)
  File "/opt/tom/github/saspy/saspy/sasdata.py", line 1102, in to_df
    return self.sas.sasdata2dataframe(self.table, self.libref, self.dsopts, method, **kwargs)
  File "/opt/tom/github/saspy/saspy/sasbase.py", line 1383, in sasdata2dataframe
    return self._io.sasdata2dataframe(table, libref, dsopts, method=method, **kwargs)
  File "/opt/tom/github/saspy/saspy/sasioiom.py", line 1588, in sasdata2dataframe
    return self.sasdata2dataframeDISK(table, libref, dsopts, **kwargs)
  File "/opt/tom/github/saspy/saspy/sasioiom.py", line 2331, in sasdata2dataframeDISK
    encoding=enc, **kwargs)
  File "/usr/lib64/python3.5/site-packages/pandas/io/parsers.py", line 702, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/usr/lib64/python3.5/site-packages/pandas/io/parsers.py", line 435, in _read
    data = parser.read(nrows)
  File "/usr/lib64/python3.5/site-packages/pandas/io/parsers.py", line 1154, in read
    df = DataFrame(col_dict, columns=columns, index=index)
  File "/usr/lib64/python3.5/site-packages/pandas/core/frame.py", line 392, in __init__
    mgr = init_dict(data, index, columns, dtype=dtype)
  File "/usr/lib64/python3.5/site-packages/pandas/core/internals/construction.py", line 212, in init_dict
    return arrays_to_mgr(arrays, data_names, index, columns, dtype=dtype)
  File "/usr/lib64/python3.5/site-packages/pandas/core/internals/construction.py", line 61, in arrays_to_mgr
    return create_block_manager_from_arrays(arrays, arr_names, axes)
  File "/usr/lib64/python3.5/site-packages/pandas/core/internals/managers.py", line 1666, in create_block_manager_from_arrays
    blocks = form_blocks(arrays, names, axes)
  File "/usr/lib64/python3.5/site-packages/pandas/core/internals/managers.py", line 1722, in form_blocks
    float_blocks = _multi_blockify(items_dict['FloatBlock'])
  File "/usr/lib64/python3.5/site-packages/pandas/core/internals/managers.py", line 1819, in _multi_blockify
    values, placement = _stack_arrays(list(tup_block), dtype)
  File "/usr/lib64/python3.5/site-packages/pandas/core/internals/managers.py", line 1861, in _stack_arrays
    stacked = np.empty(shape, dtype=dtype)
MemoryError
>>> time.localtime()
time.struct_time(tm_year=2020, tm_mon=5, tm_mday=28, tm_hour=11, tm_min=15, tm_sec=22, tm_wday=3, tm_yday=149, tm_isdst=1)
>>>
```

In this case, creating the dataframe was taking longer (presumably) than both the first 2 steps. Don't know how long it would have take if it worked. 

Well, this was much longer than I planned on, but I hope this provides you the info you need to explore this on your system. 

Let me know what you see!
Tom

**Comment by bezdomniy** [comment]
Hi @tomweber-sas , thank you for your comprehensive reply and sorry I didn't get back to you sooner. 

I think what I observed were there was a long period of low cpu and memory use must have been the first phase where the sas dataset is read and written to disk. I guess I assumed that there was some way to stream a sas7bdat file directly.

I am using IOM locally - in this case I wonder if stage 2 of to_df_DISK is necessary in this case? Can stage 1 write to a local file to be used in step 3?

Either way - your explanation of is very helpful and I will use the appropriate methods for the size of dataset I am trying to read.

Thanks,

Ilia.

**Comment by tomweber-sas** [comment]
No problem, just trying to follow up. And, yes, for the local IOM case, there is no step 2. The file written out in step 1 is the file provided to pandas in step 3, and there's no reason for a step 2 when on the same machine already. Again, you can observe that easily by providing the tempfile= and see that that file is being written in step 1, as opposed to some other file, then being transferred to that file in a step2, as would be the case for remote. 
Tom

**Comment by tomweber-sas** [comment]
Closing up old, idle issues. if you still need something on this, you can reopen it. If you need anything else, feel free to open a new issue.
Thanks!
Tom
